design_notes/XDN11 - Error handling.md

```diff
@@ -0,0 +1,194 @@
+
+## Handling language errors
+
+Producing projections may encounter a language error when projecting an API call to a component.  The error will either be from the translation of the call to the component's language or be from the component itself as a result of the call.  Either way, upon encountering an error, the projection should determine whether the error is new or is a propagation of a previous xlang error.  Strategies on doing this will be discussed in a later section.
+
+If the error is new, the projection should translate the error to an equivalent xlang error or to xlang_fail if there is no equivalent mapping.  After translating the error, the projection should call xlang's error origination API, xlang_originate_error, to store the diagnostic information collected by the language for the error.  The API will create an error info with the provided diagnostic information and store it on a thread level storage.  The created error info will remain on the thread level storage until it is retrieved by the consuming projection.
+
+xlang_originate_error will be defined in the PAL and will take the following parameters:
+- a xlang_result with the translated xlang error for the language specific error that occurred
+- a xlang_string with the language specific error that occurred
+- a xlang_string with the error message
+- a xlang_string with the error stack if available
+- a xlang_string with an identifier for the projection
+- an IUnknown pointer with any additional language specific information (can be null)
+
+```
+void xlang_originate_error(
+	xlang_result error,
+	xlang_string languageError,
+	xlang_string message,
+	xlang_string stack,
+	xlang_string projectionIdentifier,
+	IUnknown* languageInformation)
+```
+
+If the error is a propagation of a previous xlang error (i.e. the component called another xlang projected component which returned an error), then the associated error info should be retrieved along with the associated xlang error.  The retrieved xlang error should be used as the error instead of translating the language error to a new xlang error.  This will allow the identity of the error to be independent of the translations it may go through as it propagates through various languages and for it to be defined as what the initial projection which returned the error determined it was equivalent to.  Diagnostic information collected by the language should still be added to the error info as it can provide useful information in the case the error becomes a failure.  This can be done by a call to xlang_propagate_error which will add the provided information to the given error info and store it on the thread level storage.
+
+xlang_propagate_error will be defined in the PAL and will take the following parameters:
+- an IXlangErrorInfo* for the error which is being propagated
+- a xlang_string with the language specific error
```

Comment:
Nit: Consider adding \`foo\` (tics) around things that are supposed to be symbols.

Reply:
Addressed.